# Case Study: Analyzing U.S. Storm Events for Disaster Preparedness 

In this case study, we will analyze storm event data to enhance disaster preparedness efforts in the U.S. We will follow a structured data analysis workflow comprising the following phases:

1. **Ask**: Define the problem and confirm expectations.
2. **Prepare**: Collect and store data for analysis.
3. **Process**: Clean and transform data to ensure integrity.
4. **Analyze**: Use data analysis tools to draw conclusions.
5. **Share**: Interpret and communicate results to make data-driven decisions.
6. **Act**: Put insights to work to address the original problem.

We will utilize the open-source Storm Event Database provided by the National Oceanic and Atmospheric Administration (NOAA), available at [NOAA Storm Events Database](https://www.ncdc.noaa.gov/stormevents/ftp.jsp). This dataset contains comprehensive information about various storm events across the U.S., including details such as event type, location, and impact.

---

## Ask

### Problem Statement
The U.S. experiences various storm events that can impact public safety, property, and infrastructure. Analyzing the patterns and impacts of these storms will provide insights that can inform disaster preparedness and response strategies.

### Key Questions
1. **What types of storm events are most prevalent in the U.S.?**  
   Identify the most common storm events and their geographical distribution.

2. **Where do storm events most frequently occur?**  
   Map the geographical distribution of storm events to identify high-risk areas.

3. **Are certain regions more prone to specific types of storm events?**  
   Analyze the geographic hot spots for storm types to develop tailored regional preparedness plans.

4. **What is the temporal distribution of storm events?**  
   Analyze trends over time to understand seasonal patterns and changes in frequency.

5. **What is the impact of different storm events on injuries, fatalities, and property damage?**  
   Assess the severity of various storm types and their associated risks to prioritize high-risk events for preparedness planning.

6. **How can this analysis inform disaster preparedness initiatives?**  
   Determine actionable insights that can help stakeholders enhance response plans and allocate resources effectively.

---

## Prepare

### Data Sources
We will use the Storm Events Database from the NOAA for the years 2014 to 2024. The dataset can be accessed at [NOAA Storm Events Database](https://www.ncdc.noaa.gov/stormevents/ftp.jsp).

### Data Collection
The relevant files will be downloaded in CSV.GZ format, which is a compressed version of CSV files.

### Data Description
The dataset contains 51 columns with various types of information related to storm events. Below is the data dictionary presented in a table format:


| Column Name         | Example                                                    | Description                                                                                                             |
|:--------------------|:-----------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------|
| begin_yearmonth     | 201212                                                     | The year and month that the event began (%Y%m format).                                                                  |
| begin_day           | 1, 31                                                      | The day of the month that the event began (%-d format).                                                                 |
| begin_time          | 800, 2359                                                  | The time of day that the event began (%-H%M format).                                                                     |
| end_yearmonth       | 201301                                                     | The year and month that the event ended (%Y%m format).                                                                |
| end_day             | 01                                                         | The day of the month that the event ended (%-d format).                                                                  |
| end_time            | 0001                                                       | The time of day that the event ended (%-H%M format).                                                                     |
| episode_id          | 61280                                                      | ID assigned by NWS to denote the storm episode; may contain multiple events.                                            |
| event_id            | 383097                                                     | ID assigned by NWS for each individual storm event (Primary database key field).                                        |
| state               | GEORGIA                                                    | The state name where the event occurred (ALL CAPS).                                                                     |
| state_fips          | 45                                                         | Unique number assigned to the county by NIST (State FIPS).                                                              |
| year                | 2000                                                       | The four-digit year for the event in this record.                                                                       |
| month_name          | January                                                    | The name of the month for the event (not abbreviated).                                                                  |
| event_type          | Hail                                                       | The type of storm event (spelled out; not abbreviated).                                                                 |
| cz_type             | C                                                          | Indicates whether the event happened in a County/Parish, NWS Public Forecast Zone, or Marine.                           |
| cz_fips             | 245                                                        | The county FIPS number assigned by NIST or NWS Forecast Zone Number.                                                    |
| cz_name             | AIKEN                                                      | Name assigned to the county FIPS number or NWS Forecast Zone.                                                           |
| wfo                 | CAE                                                        | The NWS Forecast Office’s area of responsibility in which the event occurred.                                           |
| begin_date_time     | 04/1/2012 20:48:00                                         | Start date and time of the event (MM/DD/YYYY hh:mm:ss).                                                                 |
| cz_timezone         | EST-5                                                      | Time Zone for the County/Parish, Zone or Marine Name.                                                                   |
| end_date_time       | 04/1/2012 21:03:00                                         | End date and time of the event (MM/DD/YYYY hh:mm:ss).                                                                   |
| injuries_direct     | 1                                                          | Number of injuries directly caused by the weather event.                                                                |
| injuries_indirect   | 0                                                          | Number of injuries indirectly caused by the weather event.                                                              |
| deaths_direct       | 0                                                          | Number of deaths directly caused by the weather event.                                                                  |
| deaths_indirect     | 0                                                          | Number of deaths indirectly caused by the weather event.                                                                |
| damage_property     | 10.00K                                                     | Estimated property damage incurred by the weather event.                                                                |
| damage_crops        | 0.00K                                                      | Estimated damage to crops incurred by the weather event.                                                                |
| source              | Public                                                     | Source reporting the weather event.                                                                                     |
| magnitude           | 0.75                                                       | Measured extent of the magnitude type (only for wind speeds and hail size).                                             |
| magnitude_type      | EG                                                         | Type of magnitude measurement (e.g., wind estimated gust).                                                              |
| flood_cause         | Ice Jam                                                    | Reported cause of the flood.                                                                                            |
| category            |                                                            | Unknown (During the time of downloading this particular file, NCDC has never seen anything provided within this field.) |
| tor_f_scale         | EF0                                                        | Enhanced Fujita Scale describing tornado strength.                                                                      |
| tor_length          | 0.66                                                       | Length of the tornado while on the ground (in miles).                                                                   |
| tor_width           | 25                                                         | Width of the tornado while on the ground (in whole yards).                                                              |
| tor_other_wfo       | DDC                                                        | Continuation of a tornado segment as it crossed from one NWS Forecast Office to another.                                |
| tor_other_cz_state  | KS                                                         | Two-character representation for the state name of the continuing tornado segment.                                      |
| tor_other_cz_fips   | 41                                                         | FIPS number of the county for the continuing tornado segment.                                                           |
| tor_other_cz_name   | DICKINSON                                                  | Name of the county for the continuing tornado segment.                                                                  |
| begin_range         | 0.59                                                       | Distance to the nearest tenth of a mile to the location referenced.                                                     |
| begin_azimuth       | ENE                                                        | 16-point compass direction from the location referenced.                                                                |
| begin_location      | PINELAND                                                   | Name of city, town, or village from which the range is calculated.                                                      |
| end_range           | 0.66                                                       | See begin_range.                                                                                                        |
| end_azimuth         | WNW                                                        | See begin_azimuth.                                                                                                      |
| end_location        | RUSK                                                       | See begin_location.                                                                                                     |
| begin_lat           | 29.7898                                                    | Latitude in decimal degrees of the begin point of the event.                                                            |
| begin_lon           | -98.6406                                                   | Longitude in decimal degrees of the begin point of the event.                                                           |
| end_lat             | 29.7158                                                    | Latitude in decimal degrees of the end point of the event.                                                              |
| end_lon             | -98.7744                                                   | Longitude in decimal degrees of the end point of the event.                                                             |
| episode_narrative   | A strong upper level system over the southern Rockies...   | Narrative depicting the general nature of the episode.                                                                  |
| event_narrative     | Heavy rain caused flash flooding across parts of Wilber... | Narrative providing descriptive details of the individual event.                                                        |

### Data Storage
The data will be stored in a local environment for analysis.

In [156]:
# To begin, let's import all necessary libraries
import pandas as pd
import numpy as np
import janitor
import matplotlib.pyplot as plt
import seaborn as sns
import contextily as ctx
import glob
import missingno as msno

from shapely.geometry import Point
import geopandas as gpd
import folium

from scipy.stats import chi2_contingency

# Set display options to show all columns
pd.set_option('display.max_columns', None)

In [157]:
# In the data folder, we have lots of files. But for this analysis, we only need Storm Details files which have the prefix "StormEvents_details-ftp_v1.0_d".
# So, let's define the file path pattern
file_pattern = "./data/StormEvents_details-ftp_v1.0_d*.csv.gz"

# Get a list of all files matching the pattern
all_files = glob.glob(file_pattern)

# Read and concatenate all files into a single DataFrame
df_list = [pd.read_csv(file, compression='gzip') for file in all_files]
df_details = pd.concat(df_list, ignore_index=True)

# Rename column names to snake case for consistency 
df_details = df_details.clean_names()

# Display the first few rows of the DataFrame
df_details.head()

begin_yearmonth  begin_day  begin_time  end_yearmonth  end_day  end_time  \
0           202112         11         349         202112       11       350   
1           202112         11         249         202112       11       254   
2           202112         11         325         202112       11       327   
3           202112         11         232         202112       11       239   
4           202112          6         724         202112        6       724   

   episode_id  event_id      state  state_fips  year month_name  \
0      165322    999750  TENNESSEE          47  2021   December   
1      165322    999613  TENNESSEE          47  2021   December   
2      165322    999636  TENNESSEE          47  2021   December   
3      165322    999604  TENNESSEE          47  2021   December   
4      165321    999306  TENNESSEE          47  2021   December   

          event_type cz_type  cz_fips   cz_name  wfo     begin_date_time  \
0            Tornado       C      165    SUMNER  OHX  11-DEC-21 03:49:00   
1            Tornado       C       43   DICKSON  OHX  11-DEC-21 02:49:00   
2  Thunderstorm Wind       C       37  DAVIDSON  OHX  11-DEC-21 03:25:00   
3            Tornado       C       81   HICKMAN  OHX  11-DEC-21 02:32:00   
4  Thunderstorm Wind       C       49  FENTRESS  OHX  06-DEC-21 07:24:00   

  cz_timezone       end_date_time  injuries_direct  injuries_indirect  \
0       CST-6  11-DEC-21 03:50:00                0                  0   
1       CST-6  11-DEC-21 02:54:00                0                  0   
2       CST-6  11-DEC-21 03:27:00                0                  0   
3       CST-6  11-DEC-21 02:39:00                0                  0   
4       CST-6  06-DEC-21 07:24:00                0                  0   

   deaths_direct  deaths_indirect damage_property damage_crops  \
0              0                0          10.00K        0.00K   
1              0                0          10.00K        0.00K   
2              0                0         250.00K        0.00K   
3              0                0          50.00K        0.00K   
4              0                0           3.00K        0.00K   

             source  magnitude magnitude_type flood_cause  category  \
0  NWS Storm Survey        NaN            NaN         NaN       NaN   
1  NWS Storm Survey        NaN            NaN         NaN       NaN   
2  NWS Storm Survey       74.0             EG         NaN       NaN   
3  NWS Storm Survey        NaN            NaN         NaN       NaN   
4      Social Media       52.0             EG         NaN       NaN   

  tor_f_scale  tor_length  tor_width tor_other_wfo tor_other_cz_state  \
0         EF0        1.72       50.0           OHX                 TN   
1         EF0        5.41      175.0           NaN                NaN   
2         NaN         NaN        NaN           NaN                NaN   
3         EF1        8.54      400.0           NaN                NaN   
4         NaN         NaN        NaN           NaN                NaN   

   tor_other_cz_fips tor_other_cz_name  begin_range begin_azimuth  \
0              189.0            WILSON          3.0           WNW   
1                NaN               NaN          1.0           ENE   
2                NaN               NaN          1.0            NW   
3                NaN               NaN          4.0            NW   
4                NaN               NaN          1.0             W   

  begin_location  end_range end_azimuth end_location  begin_lat  begin_lon  \
0     HUNTERS PT        3.0          NW   HUNTERS PT    36.3178   -86.3235   
1        TIDWELL        2.0         ESE  BAKERSWORKS    36.0255   -87.3054   
2      MAPLEWOOD        2.0          SW        AMQUI    36.2372   -86.7286   
3           SPOT        4.0         NNW     PINEWOOD    35.9205   -87.6423   
4      JAMESTOWN        1.0           W    JAMESTOWN    36.4322   -84.9405   

   end_lat  end_lon                                  episode_narrative  \
0  36.32

---

## Process

In [158]:
# Display the DataFrame information
df_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 669746 entries, 0 to 669745
Data columns (total 51 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   begin_yearmonth     669746 non-null  int64  
 1   begin_day           669746 non-null  int64  
 2   begin_time          669746 non-null  int64  
 3   end_yearmonth       669746 non-null  int64  
 4   end_day             669746 non-null  int64  
 5   end_time            669746 non-null  int64  
 6   episode_id          669746 non-null  int64  
 7   event_id            669746 non-null  int64  
 8   state               669746 non-null  object 
 9   state_fips          669746 non-null  int64  
 10  year                669746 non-null  int64  
 11  month_name          669746 non-null  object 
 12  event_type          669746 non-null  object 
 13  cz_type             669746 non-null  object 
 14  cz_fips             669746 non-null  int64  
 15  cz_name             669746 non-nul

In [159]:
# Based on the data dictionary, we can drop the following columns: `category`
df_details.drop(columns=['category'], inplace=True)

In [160]:
# Convert `begin_date_time` and `end_date_time` columns to datetime type. Current data follows the format 11-DEC-21 03:49:00
df_details['begin_date_time'] = pd.to_datetime(df_details['begin_date_time'], format='%d-%b-%y %H:%M:%S')
df_details['end_date_time'] = pd.to_datetime(df_details['end_date_time'], format='%d-%b-%y %H:%M:%S')

# Convert `tor_other_cz_fips` to integer type
df_details['tor_other_cz_fips'] = df_details['tor_other_cz_fips'].astype('Int64')

# Convert `state`, `month_name`, `event_type`, `cz_type`, `wfo`, `cz_timezone`, `magnitude_type`, `tof_f_scale`, `tor_other_wfo`, `tor_other_cz_state` to categorical type
df_details['state'] = df_details['state'].astype('category')
df_details['month_name'] = df_details['month_name'].astype('category')
df_details['event_type'] = df_details['event_type'].astype('category')
df_details['cz_type'] = df_details['cz_type'].astype('category')
df_details['cz_timezone'] = df_details['cz_timezone'].astype('category')
df_details['magnitude_type'] = df_details['magnitude_type'].astype('category')
df_details['tor_f_scale'] = df_details['tor_f_scale'].astype('category')
df_details['tor_other_cz_state'] = df_details['tor_other_cz_state'].astype('category')

In [161]:
# Check the data information
df_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 669746 entries, 0 to 669745
Data columns (total 50 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   begin_yearmonth     669746 non-null  int64         
 1   begin_day           669746 non-null  int64         
 2   begin_time          669746 non-null  int64         
 3   end_yearmonth       669746 non-null  int64         
 4   end_day             669746 non-null  int64         
 5   end_time            669746 non-null  int64         
 6   episode_id          669746 non-null  int64         
 7   event_id            669746 non-null  int64         
 8   state               669746 non-null  category      
 9   state_fips          669746 non-null  int64         
 10  year                669746 non-null  int64         
 11  month_name          669746 non-null  category      
 12  event_type          669746 non-null  category      
 13  cz_type             669746 no

In [162]:
# Now let check for duplicates in the data
duplicates = df_details.duplicated().sum()
print(f"Number of duplicate rows: {duplicates}")

Number of duplicate rows: 0


That's great news. We have no duplicated rows in our data. Let's proceed to check for missing values in the data.

In [163]:
# Check for missing values in the data
missing_values = df_details.isnull().sum()
print("Missing values in the data:")
missing_values[missing_values > 0].sort_values(ascending=False)

Missing values in the data:


tor_other_cz_name     667758
tor_other_wfo         667758
tor_other_cz_state    667758
tor_other_cz_fips     667758
tor_f_scale           654554
tor_length            654554
tor_width             654554
flood_cause           597354
magnitude_type        416845
magnitude             320801
end_lat               259317
end_lon               259317
begin_range           259317
begin_azimuth         259317
begin_location        259317
end_range             259317
end_azimuth           259317
end_location          259317
begin_lat             259317
begin_lon             259317
event_narrative       140063
damage_property       136418
damage_crops          134127
dtype: int64

We have several columns with missing values. Let's check for the columns with missing values.


Start with `tor_f_scale`, `tor_length`, and `tor_width` columns.
Based on the Data Dictionary provided at the beginning I assume columns like `tor_f_scale`, `tor_length`, and `tor_width` are relevant only for tornado events. It’s expected that these fields would be missing for non-tornado events. Let's prove it.

In [180]:
# Select the tornado-specific columns
tornado_columns = ['tor_f_scale', 'tor_length', 'tor_width']

# Check if any of these columns is populated for `event_type` not containing 'Tornado'
non_tornado = df_details[~df_details['event_type'].str.contains('Tornado', na=False)][tornado_columns].notnull().any(axis=1).sum()

# Check if these columns are NOT all null for event_type containing 'Tornado'
tornado= df_details[df_details['event_type'].str.contains('Tornado', na=False)][tornado_columns].isnull().all(axis=1).sum()

# Output the results
print(f"Number of non-tornado events with tornado-specific columns filled: {non_tornado}")
print(f"Number of tornado events with missing tornado-specific columns: {tornado}")

Number of non-tornado events with tornado-specific columns filled: 0
Number of tornado events with missing tornado-specific columns: 0


The result indicates that there are no missing values in the tornado-specific columns (`tor_f_scale`, `tor_length`, and `tor_width`) for events classified as tornadoes in the dataset. This confirms my assumption that these fields are only populated for tornado events and should not have missing values within that context.

Thus, we don’t need to worry about these missing values in the broader dataset, as they are valid and expected for non-tornado events.

Now, let's move on to `tor_other_cz_name`, `tor_other_wfo`, `tor_other_cz_state`, and `tor_other_cz_fips` columns. Based on the Data Dictionary mentioned above, columns like `tor_other_cz_name`, `tor_other_wfo`, `tor_other_cz_state`, and `tor_other_cz_fips` apply only to tornadoes that cross into other geographical areas. If a tornado travels beyond its initial location, these additional columns should contain data.


In [181]:
# Define the columns related to tornadoes that cross geographical areas
tornado_other_location_cols = ['tor_other_cz_name', 'tor_other_wfo', 'tor_other_cz_state', 'tor_other_cz_fips']

# 1. Prove there are only two outcomes: either all four are null or all four are not null
# Create a mask where either all 4 are null or all 4 are not null
all_null_or_not_null = (
    (df_details[tornado_other_location_cols].isnull().all(axis=1)) | 
    (df_details[tornado_other_location_cols].notnull().all(axis=1))
)

# Check if any rows violate this condition
invalid_rows = df_details[~all_null_or_not_null]

if invalid_rows.empty:
    print("There are no cases where only some of the 'tor_other_*' columns are null/non-null. The assumption holds.")
else:
    print("There are cases where only some of the 'tor_other_*' columns are null/non-null:")
    print(invalid_rows)

# 2. Prove that if the 4 columns are null, event_type is not Tornado, and if not null, event_type is Tornado and the original locations differs from the other locations

# First check if event_type is not Tornado when all 4 columns are null
non_tornado_mismatch = df_details[(df_details[tornado_other_location_cols].notnull().all(axis=1)) & (~df_details['event_type'].str.contains('Tornado', na=False))]

# Check if the initial locations are different from the other locations when the 4 columns are not null and event_type is Tornado
location_mismatch = df_details[(df_details[tornado_other_location_cols].notnull().all(axis=1)) & 
                               (df_details['event_type'].str.contains('Tornado', na=False)) &
                               (df_details['cz_fips'] == df_details['tor_other_cz_fips']) &
                                (df_details['cz_name'] == df_details['tor_other_cz_name']) &
                                (df_details['wfo'] == df_details['tor_other_wfo'])]

# Output the results
if non_tornado_mismatch.empty and location_mismatch.empty:
    print("All conditions hold: tornado events have 'tor_other_*' columns populated and locations differ when these columns are populated.")
else:
    if not non_tornado_mismatch.empty:
        print("There are non-tornado events where 'tor_other_*' columns are populated:")
        print(non_tornado_mismatch[['event_type'] + tornado_other_location_cols])
    if not location_mismatch.empty:
        print("There are tornado events where the initial location equals the other location but 'tor_other_*' columns are populated:")
        print(location_mismatch[['event_id', 'event_type', 'cz_name', 'wfo', 'cz_type', 'cz_fips',] + tornado_other_location_cols])

There are no cases where only some of the 'tor_other_*' columns are null/non-null. The assumption holds.
All conditions hold: tornado events have 'tor_other_*' columns populated and locations differ when these columns are populated.


This means that these columns only contain data when a tornado crosses into a new geographical area. If a tornado stays within one boundary, these columns remain null. This is consistent with the data dictionary and the expected behavior of the dataset.


Now, we can move on to the other columns related to geographical locations (`end_lat`, `end_lon`, `begin_range`, `begin_azimuth`, `begin_location`, `end_range`, `end_azimuth`, `end_location`, `begin_lat`, `begin_lon`). I noticed that they all have the same number of rows with missing values. So, I suspect either all have values or are all null in a row, with no partial cases where some columns are populated while others are not. Let's verify this assumption.

In [184]:
# Select the relevant columns
geo_columns = ['end_lat', 'end_lon', 'begin_range', 'begin_azimuth', 'begin_location', 
               'end_range', 'end_azimuth', 'end_location', 'begin_lat', 'begin_lon']

# Create a subset of just those columns
geo_data = df_details[geo_columns]

# Check if all columns are either fully null or fully populated
all_null_or_all_filled = geo_data.isnull().all(axis=1) | geo_data.notnull().all(axis=1)

# Count the number of rows where some columns are null but others are not
partially_null_rows_count = (~all_null_or_all_filled).sum()

print(f"Number of rows where some columns are null and others are filled: {partially_null_rows_count}")

Number of rows where some columns are null and others are filled: 0


So, that confirms my theory. These columns are either all being populated or all null in a row, with no partial cases where some columns are populated while others are not. This leads to me to believe that these columns are not always required for every event in the dataset. It makes sense that events like droughts or heat waves, for example, may not have clearly defined start or end geographical locations because they can affect broad regions over time rather than specific points.

We can check if the missing geographical data is associated with specific types of events that typically don’t have a clear starting or ending location. 

In [185]:
# Filter rows where the geographical data is missing
missing_geo_data = df_details[geo_data.isnull().all(axis=1)]

# Count the event types for these rows
missing_geo_event_types = missing_geo_data['event_type'].value_counts()

print("Event types associated with missing geographical data:")
missing_geo_event_types

Event types associated with missing geographical data:


event_type
Winter Weather                40129
High Wind                     38392
Winter Storm                  31873
Drought                       31702
Heavy Snow                    24229
Heat                          13209
Excessive Heat                11608
Strong Wind                   11197
Extreme Cold/Wind Chill        9508
Dense Fog                      7529
Cold/Wind Chill                6295
Frost/Freeze                   6087
Blizzard                       5888
High Surf                      4326
Wildfire                       3447
Tropical Storm                 2664
Ice Storm                      2566
Coastal Flood                  2385
Lake-Effect Snow               1028
Dust Storm                      982
Rip Current                     868
Marine Tropical Storm           506
Storm Surge/Tide                443
Astronomical Low Tide           353
Avalanche                       348
Sleet                           321
Lakeshore Flood                 315
Hurricane        

This result supports the idea that the missing geographical data (latitude, longitude, range, azimuth, and location) is associated with events that typically don’t have well-defined start and end points.

Given this context, we can conclude that the missing values in these columns are valid and do not require further action.

Next, let's check the `flood_cause` column. The `flood_cause` column is expected to be populated only for events classified as floods. Let's verify this assumption.

In [190]:
# Check if any of the flood events does not have the flood_cause column populated
flood = df_details[df_details['event_type'].str.contains('Flood', na=False)]['flood_cause'].isnull().sum()

# Check if any of the non-flood events have the flood_cause column populated
non_flood = df_details[~df_details['event_type'].str.contains('Flood', na=False)]['flood_cause'].notnull().sum()

# Display the result
print(f"Number of flood events with missing 'flood_cause': {flood}")
print(f"Number of non-flood events with 'flood_cause' filled: {non_flood}")

Number of flood events with missing 'flood_cause': 2700
Number of non-flood events with 'flood_cause' filled: 1409


Let's check what `event_type` values are associated with these exceptions.

In [195]:
# Display the event types for flood events with missing flood_cause
flood_missing_cause = df_details[df_details['event_type'].str.contains('Flood', na=False) & df_details['flood_cause'].isnull()]['event_type'].value_counts()

# Output the results
flood_missing_cause[flood_missing_cause > 0]

event_type
Coastal Flood      2385
Lakeshore Flood     315
Name: count, dtype: int64

In [196]:
# Display the event types for non-flood events with flood_cause filled. Only display rows with count > 0
non_flood_cause = df_details[~df_details['event_type'].str.contains('Flood', na=False) & df_details['flood_cause'].notnull()]['event_type'].value_counts()

# Output the results
non_flood_cause[non_flood_cause > 0]


event_type
Debris Flow    1409
Name: count, dtype: int64

The result shows that the `flood_cause` column is filled for events that are not classified as floods. In this case, 1409 rows have the `flood_cause` column populated while their associated `event_type` is "Debris Flow".

Debris flow can be influenced by several factors, but it typically occurs as a result of **heavy rainfall**. This could explain why the `flood_cause` column is populated for these events. So, we do not need to handle these missing values in the `flood_cause` column as they are valid entries.

Now, let's check the `magnitude` column. The `magnitude` column is expected to be populated for events classified as wind or hail. Let's verify this assumption.

In [ ]:
# Check for missing values in the magnitude column
missing_magnitude = df_details['magnitude'].isnull().sum()
print(f"Number of missing values in 'magnitude': {missing_magnitude}\n")

# Identify event types with missing magnitude
missing_magnitude_events = df_details[df_details['magnitude'].isnull()]['event_type'].value_counts()
print("Event types with missing magnitude:")
missing_magnitude_events

In [ ]:
# Check non-null magnitude values and their associated event types
non_null_magnitude_events = df_details[df_details['magnitude'].notnull()]['event_type'].value_counts()
print("Event types with non-null magnitude:")
non_null_magnitude_events

The majority of the event types associated with missing `magnitude` values are not typically associated with specific magnitudes, such as "Flash Flood", "Winter Weather", "Winter Storm" and other while the event types that have non-null values for `magnitude` primarily include "Thunderstorm Wind", "Hail", and "High Wind". This aligns with the definition provided in the Data Dictionary. Thus, the missing values in the `magnitude` column are expected and do not require any further action.

Now, let's check `magnitude_type` column. The `magnitude_type` column is expected to be populated for events classified as wind and has an associated value in the `magnitude` column. Let's verify this assumption.

In [ ]:
# Filter rows where magnitude_type is not null
magnitude_type_non_null = df_details[df_details['magnitude_type'].notnull()]

# Check if magnitude_type rows have non-null values in magnitude (Condition 1)
magnitude_missing = magnitude_type_non_null[magnitude_type_non_null['magnitude'].isnull()]

# Define wind-related event types (Condition 2)
wind_event_types = ['Thunderstorm Wind', 'High Wind', 'Strong Wind', 'Marine Thunderstorm Wind', 
                    'Marine High Wind', 'Marine Strong Wind']

# Check if magnitude_type rows have wind-related event types (Condition 2)
non_wind_events = magnitude_type_non_null[~magnitude_type_non_null['event_type'].isin(wind_event_types)]

# Output results
print(f"Rows where magnitude_type is non-null but magnitude is null: {len(magnitude_missing)}")
print(f"Event types with non-null magnitude_type but not wind-related: {len(non_wind_events)}")

# If any violations are found, display the problematic rows
if not magnitude_missing.empty:
    print("Rows with magnitude_type but missing magnitude:")
    print(magnitude_missing[['event_type', 'magnitude_type', 'magnitude']])

if not non_wind_events.empty:
    print("Rows with magnitude_type but non-wind event type:")
    print(non_wind_events[['event_type', 'magnitude_type']])

The `magnitude_type` field is correctly populated only for wind-related events and always has an associated value in the `magnitude` column. No cases were found where `magnitude_type` was non-null without a corresponding value in `magnitude`, or where `magnitude_type` was filled for non-wind events. This confirms that the field is used appropriately in the dataset. So no further action is required.

Now, let's check `event_narrative` column. The `event_narrative` column provides descriptive details of the individual event. Let's verify if the missing values in this column are associated with specific event types.

In [ ]:
# Check for missing values in the event_narrative column
missing_narrative_events = df_details[df_details['event_narrative'].isnull()]

# Count the event types in these rows
event_narrative_counts = missing_narrative_events['event_type'].value_counts()

print("Event types with missing event_narrative:")
event_narrative_counts

In our analysis, we observed a significant number of missing values in the `event_narrative` column. Given that the narrative primarily provides additional context rather than contributing to the quantitative analysis, we will not be filling in these missing values. Instead, we will retain the existing data as is, ensuring that our analysis remains focused on quantifiable metrics. Furthermore, since the `episode_narrative` column has no missing values, it may serve as a useful supplementary reference, although it will not be included in our main analysis.



For the `damage_property` column, we will impute the missing values, which account for approximately 20.38% of the dataset. Given this high percentage, removing these missing rows could significantly bias our analysis and lead to a loss of valuable information. We will use averages calculated for combinations of `state`, `event_type`, and `year` for most cases, but where all values in a group are missing, we will resort to 0. This mixed approach allows us to consider regional and event-specific variations while still ensuring that we have a method for imputing values when necessary. By leveraging both types of averages, we aim to provide a more accurate estimate than relying solely on an overall average.

In [ ]:
# Convert 'damage_property' to numeric
def convert_damage(value):
    if isinstance(value, str):
        if 'K' in value:
            return float(value.replace('K', '').strip()) * 1_000
        elif 'M' in value:
            return float(value.replace('M', '').strip()) * 1_000_000
        elif 'B' in value:
            return float(value.replace('B', '').strip()) * 1_000_000_000
    return np.nan

# Apply the conversion
df_details['damage_property_numeric'] = df_details['damage_property'].apply(convert_damage)

# Calculate average damage_property_numeric by state, event_type, and year
averages = df_details.groupby(['state', 'event_type', 'year'])['damage_property_numeric'].mean().reset_index()

# Rename the column for clarity
averages.rename(columns={'damage_property_numeric': 'average_damage_property'}, inplace=True)

# Merge the averages back to the original DataFrame
df_details = df_details.merge(averages, on=['state', 'event_type', 'year'], how='left')

# Identify the groups where all values in a group are missing
group_missing = df_details.groupby(['state', 'event_type', 'year'])['damage_property_numeric'].transform(lambda x: x.isna().all())

# Impute missing values with the calculated averages
df_details['imputed_damage_property'] = df_details['damage_property_numeric'].fillna(df_details['average_damage_property'])

# For groups where all values are missing, impute with 0
df_details.loc[group_missing, 'imputed_damage_property'] = df_details.loc[group_missing, 'imputed_damage_property'].fillna(0)

# Check for remaining missing values
remaining_missing = df_details['imputed_damage_property'].isnull().sum()
print(f"Remaining missing values in 'damage_property': {remaining_missing}")

# Drop the temporary columns
df_details.drop(columns=['average_damage_property'], inplace=True)

Similar for `damage_crops` column, we will impute the missing values. We will follow the same approach as we did for the `damage_property` column.

In [ ]:
# Convert 'damage_crops' to numeric
df_details['damage_crops_numeric'] = df_details['damage_crops'].apply(convert_damage)

# Calculate average damage_crops_numeric by state, event_type, and year
averages_crops = df_details.groupby(['state', 'event_type', 'year'])['damage_crops_numeric'].mean().reset_index()

# Rename the column for clarity
averages_crops.rename(columns={'damage_crops_numeric': 'average_damage_crops'}, inplace=True)

# Merge the averages back to the original DataFrame
df_details = df_details.merge(averages_crops, on=['state', 'event_type', 'year'], how='left')

# Identify the groups where all values in a group are missing
group_missing_crops = df_details.groupby(['state', 'event_type', 'year'])['damage_crops_numeric'].transform(lambda x: x.isna().all())

# Impute missing values with the calculated averages
df_details['imputed_damage_crops'] = df_details['damage_crops_numeric'].fillna(df_details['average_damage_crops'])

# For groups where all values are missing, impute with 0
df_details.loc[group_missing_crops, 'imputed_damage_crops'] = df_details.loc[group_missing_crops, 'imputed_damage_crops'].fillna(0)

# Check for remaining missing values
remaining_missing_crops = df_details['imputed_damage_crops'].isnull().sum()
print(f"Remaining missing values in 'damage_crops': {remaining_missing_crops}")

# Drop the temporary columns
df_details.drop(columns=['average_damage_crops'], inplace=True)

## Appendix

In [ ]:
# Calculate percentage of missing values by state
missing_by_state = df_details['damage_property'].isnull().groupby(df_details['state']).mean() * 100

missing_by_state.sort_values(ascending=False)

In [ ]:
# Calculate percentage of missing values by event_type
missing_by_event_type = df_details['damage_property'].isnull().groupby(df_details['event_type']).mean() * 100

missing_by_event_type.sort_values(ascending=False)

In [ ]:
# Calculate percentage of missing values by year
missing_by_year = df_details['damage_property'].isnull().groupby(df_details['year']).mean() * 100

missing_by_year.sort_values(ascending=False)

In [ ]:
# Group by year, state, and event_type, then calculate the percentage of missing damage_property
missing_groups = df_details.groupby(['year', 'state', 'event_type'])['damage_property'].apply(lambda x: x.isna().mean()).reset_index()

# Filter for groups where the missing percentage is 100%
missing_100_percent = missing_groups[missing_groups['damage_property'] == 1.0]

# Display the result
missing_100_percent

### Outline for Python Notebook: Storm Event Data Analysis

#### 1. **Ask**
   - **Define the Problem**
     - Identify the key questions: 
       - What types of storms are most common in the US?
       - Where do most storms originate?
       - How do storms impact communities (injuries, deaths, damage)?
   - **Confirm Stakeholder Expectations**
     - Define what stakeholders expect from the analysis (e.g., actionable insights, visualizations).

#### 2. **Prepare**
   - **Collect Data**
     - Load the storm event dataset using `pandas`.
     - Explore the dataset structure using `info()`, `head()`, and `describe()`.
   - **Store Data**
     - Save any preliminary data transformations or filtered datasets for later use.

#### 3. **Process**
   - **Clean the Data**
     - Handle missing values and outliers.
     - Convert data types as necessary (e.g., date parsing).
   - **Transform Data**
     - Standardize column names to snake_case.
     - Filter data for specific storm events or time periods if needed.
     - Create new columns for analysis (e.g., total damage).

#### 4. **Analyze**
   - **Descriptive Analysis**
     - Count unique storm events by type.
     - Calculate total injuries, deaths, and property damage.
   - **Geospatial Analysis**
     - Use latitude and longitude data to plot storm origins on a map.
     - Identify regions most affected by specific storm types.
   - **Trends and Patterns**
     - Analyze trends over time (e.g., increase in storm frequency).

#### 5. **Share**
   - **Visualizations**
     - Create graphs and maps to illustrate key findings.
     - Use interactive maps for better engagement.
   - **Interpret Results**
     - Summarize insights from the data (e.g., most common storms, geographic hotspots).

#### 6. **Act**
   - **Recommendations**
     - Provide actionable insights based on findings.
     - Suggest further areas for research or monitoring based on trends.

### Notes
- Ensure to document each step with comments and markdown cells to explain your thought process and findings.
- Include visualizations at appropriate points to enhance understanding.
